### ISSUE TITLE
Consider deprecating and/or modifying behavior of std::env::set_var

### ISSUE THREAD

**Issue by leo60228** [rust-lang/rust#90308]
Observing an environment variable concurrently with a call to `setenv` constitutes a data race. Rust currently handles this by using a mutex, but that only protects against functions in `std::env`. My interpretation of POSIX (which appears to be the same as the glibc developers') is that any libc function is allowed to call `getenv`. This has already caused problems with `getaddrinfo` in #27970.
Additionally, a large amount of C code calls `getenv` but is otherwise thread-safe. While this isn't necessarily the standard library's issue, it's impossible for third-party libraries to soundly use this code without requiring the user to certify that this isn't an issue via an `unsafe` block. Some examples of this happening in practice are in https://github.com/time-rs/time/issues/293 and https://github.com/rust-lang/flate2-rs/issues/272.
https://github.com/rustsec/advisory-db/issues/926 had several proposals on how this could be handled brought up.

## Make `std::env::set_var` unsafe
This is arguably the cleanest solution to the issue. This could be considered a soundness fix, which would make it an option, but the ecosystem impact feels like it'd be too big.

## Don't actually call `setenv`
`std` could keep track of the environment itself, and make `set_var` changes only visible to `var` and subprocesses. This is probably the way to solve the issue with the least impact on existing code, but the behavior is somewhat unexpected and not zero-cost.

## Only call `setenv` in single-threaded code
This would reduce the impact further, but seems even less expected for negligible benefit.

## Deprecate `std::env::set_var`
This would make it clear that setting an environment variable in the current process is discouraged. It could also be combined with not actually calling `setenv`, which would be my preferred solution to this issue.

**Comment by leo60228** [comment]
For reference, .NET takes option #2.

**Comment by cgwalters** [comment]
Hi, there's a bunch of discussion on this issue, but I think https://internals.rust-lang.org/t/synchronized-ffi-access-to-posix-environment-variable-functions/15475 has the most information right now.

**Comment by briansmith** [comment]
> Hi, there's a bunch of discussion on this issue, but I think https://internals.rust-lang.org/t/synchronized-ffi-access-to-posix-environment-variable-functions/15475 has the most information right now.

There is a lot of discussion in #27970 that's really about this issue. This is largely my fault because I didn't know this issue existed.

**Comment by briansmith** [comment]
I looked at an application that is using `set_var` in multiple places:
* To provide a default value for `RUST_LOG=debug` when it wasn't already set. We'd prefer a way to do this without setting the environment variable.
* To force `RUST_BACKTRACE=full` regardless of what the environment says. Again, we'd prefer a way to do this without setting an environment variable.
* To test functions that read from the environment. At some point we should have a new solution for supporting this.
* To do shell scripting, but in Rust. Not a big deal for us to use an `unsafe` function for this, though not ideal.

**Comment by Nemo157** [comment]
> * To provide a default value for `RUST_LOG=debug` when it wasn't already set. We'd prefer a way to do this without setting the environment variable.

https://docs.rs/env_logger/latest/env_logger/struct.Env.html#method.filter_or

> * To force `RUST_BACKTRACE=full` regardless of what the environment says. Again, we'd prefer a way to do this without setting an environment variable.

https://github.com/rust-lang/rust/pull/79085

**Comment by leo60228** [comment]
> To test functions that read from the environment. At some point we should
have a new solution for supporting this.

I don't think there's a reasonable way to support this other than running
that test in its own process. That's already prone to logic race conditions
since tests are multithreaded.

At the very least, in an integration test it should be pretty much
obviously safe to put the relevant `set_env` calls at the top of `main`.

**Comment by briansmith** [comment]
First, it turns out that our use of `RUST_LOG` was a bit different than I described. I agree current APIs are sufficient, at least for the logging crates we use.

As far as testing goes, if the `serial_test` crate truly ensures no other threads are doing anything during the test, then that might work for our test uses of `set_var` and `remove_var`.

**Comment by briansmith** [comment]
In order to use `serial_test` to make `set_var` and `remove_var` in a test safe, the tests would have to be integration tests and where there are no other tests in the same file (really, resultant binary), so that all the tests in the binary are marked `#[serial]`, because `#[serial]` from the `serial_test` crate only serializes `#[serial]` tests; it doesn't prevent non-`#[serial]` tests from executing concurrently with `#[serial]` tests. Further, all of those `#[serial]` tests would need to avoid spawning threads (that outlive the test). So it'd be pretty error-prone.

**Comment by RalfJung** [comment]
> To provide a default value for RUST_LOG=debug when it wasn't already set. We'd prefer a way to do this without setting the environment variable.

Miri needs something similar [here](https://github.com/rust-lang/miri/blob/b258eb3c27a87320c35860dc233a656838f65003/src/bin/miri.rs#L169), and I think we somehow have two loggers here in this process so I don't know if we can use the env_logger API as a replacement.

**Comment by cgwalters** [comment]
>  Further, all of those #[serial] tests would need to avoid spawning threads (that outlive the test).

And even inside such a serial test, any use of e.g. tokio or rayon could spawn threads in a not immediately obvious way.  So I wouldn't bet on this approach as a good mitigation.

**Comment by briansmith** [comment]
@cgwalters wrote:
> And even inside such a serial test, any use of e.g. tokio or rayon could spawn threads in a not immediately obvious way. So I wouldn't bet on this approach as a good mitigation.

That's right. For the tests we have, what I suggested above would work. What we're more likely to do is refactor the code so that reading the environment is done separately from the functionality being tested, and then run the tests using a private API that allows us to pass in a fake environment stored in a `HashMap` or similar. So we have workarounds that are good enough for a lot of applications, and I don't think fully solving the testing issue should block the deprecation, as long as an `unsafe` variant is added during the deprecation.

**Comment by briansmith** [comment]
> Miri needs something similar [here](https://github.com/rust-lang/miri/blob/b258eb3c27a87320c35860dc233a656838f65003/src/bin/miri.rs#L169), and I think we somehow have two loggers here in this process so I don't know if we can use the env_logger API as a replacement.

It seems like that could be solved by just making `rustc_driver`'s log initialization API more flexible, and then taking advantage of that flexibility?

**Comment by briansmith** [comment]
It seems like thie current plan is to to mark these APIs as `#[deprecated_safe]`, though I'm making this statement based on inference.

**Comment by jhpratt** [comment]
Correct, but the PR for that was closed after the author deleted their GitHub account. I have the branch archived on my fork.

**Comment by thomcc** [comment]
Is anybody working on a #[deprecated_safe] (or `#[rustc_deprecated_safe]` more likely)?

**Comment by vi** [comment]
Will it the `#[deprecated_safe]` also be applied to [`CommandExt::before_exec`](https://doc.rust-lang.org/stable/std/os/unix/process/trait.CommandExt.html#method.before_exec)?

> `#[rustc_deprecated_safe]` more likely

`#[deprecated]` was once unstable too, but is now available for everybody. Why should `#[deprecated_safe]` be inherently unstable?

**Comment by RalfJung** [comment]
So what about calls to `set_env` that happen to already be in an `unsafe`, and thus their authors would not even notice the `deprecated_safe`? I wonder if there is a good way to figure out how common this pattern is.

**Comment by jhpratt** [comment]
In theory, we could add a hack to the compiler so that it errors when the method is encountered in an unsafe context. That could then be run on crater. Unless I'm missing something?

**Comment by RalfJung** [comment]
Yeah that's about the best plan I can come up with, too.

**Comment by jhpratt** [comment]
I'll attempt an implementation for that later today, hopefully.

**Comment by jhpratt** [comment]
PR for the experiment is up: #103752.

**Comment by briansmith** [comment]
> So what about calls to `set_env` that happen to already be in an `unsafe`, and thus their authors would not even notice the `deprecated_safe`? I wonder if there is a good way to figure out how common this pattern is.

Consider the user who read these conversations and decided to put all their calls into `unsafe { ... }` blocks preemptively, before the functions are marked unsafe in the library. Their code shouldn't break as they're doing the right thing. So I think PR #103752 makes sense as far as doing a crater run to measure how common it is, but it shouldn't be merged into a (nightly) release.

**Comment by jhpratt** [comment]
@briansmith It's not intended to be merged, even if it returns no breakage. The only purpose is to measure the _potential_ issues.

**Comment by RalfJung** [comment]
https://github.com/rust-lang/rust/pull/103752 found a small but non-zero number of crates that use set_env / remove_env is an unsafe block and thus would get no warning from a `deprecated_safe`.

I think at this point we have all the data and need to make a decision. That will certainly involve @rust-lang/libs-api and might require an RFC?

**Comment by RalfJung** [comment]
Nominating for T-libs-api discussion. There's a lot of background material here, which I am not sure how to best summarize:
- This issue
- https://github.com/rust-lang/rust/issues/27970
- A [discussion on IRLO](https://internals.rust-lang.org/t/synchronized-ffi-access-to-posix-environment-variable-functions/15475)
- https://github.com/rustsec/advisory-db/issues/926

The issue OP describes some of the options that were considered:
1. deprecate the safety of `set_var` (and `remove_var`)
2. deprecate `set_var` outright and introduce new, unsafe alternatives
3. make `set_var` panic if there is more than 1 thread (offering an `unsafe` alternative that does not panic)

I am not listing options that involve `set_var` not actually mutating the process environment, as those seem to be a non-started based on some T-libs feedback in the past. https://github.com/rust-lang/rust/pull/103752 gathered some data to help decide between the first 2 options -- the problem with the first option being that calls to `set_var` that are already in an unsafe block will not get any warning.

Besides that main choice, there are some other decisions that also need to be made:
- Should we get rid of the env lock entirely, or is it still helpful? It doesn't provide soundness, but it might be able to mitigate some incorrect concurrent uses of `set_var`. Getting rid of it entirely could make it easier to resolve https://github.com/rust-lang/rust/issues/64718 but I am not sure about that.
- Should we provide some new API for *pure Rust code* to migrate to? Rust could have its own environment that is separate from the process environment and affects `Command` and `var`. If we don't do this, then what else do we tell existing users of `set_var` in concurrent programs that don't actually care about the process environment, they use use `env::var`/`env::set_var` as an easy-to-access bit of global state within a Rust program?

**Comment by m-ou-se** [comment]
Thanks for the summary, @RalfJung! We discussed this in the @rust-lang/libs-api meeting. We like option 1 best.


> Should we get rid of the env lock entirely, or is it still helpful? It doesn't provide soundness, and it makes the library more complicated, but it might be able to mitigate some incorrect concurrent uses of set_var.

We don't think we can remove the lock, because that could make existing sound (pure Rust) code unsound.

> Should we provide some new safe API for pure Rust code to migrate to? Rust could have its own environment that is separate from the process environment and affects Command and var. If we don't do this, then what else do we tell existing users of set_var in concurrent programs that don't actually care about the process environment; they just use env::var/env::set_var as an easy-to-access bit of global state within a Rust program?

A new safe API that works on some kind of separate Rust only copy of the environment would probably introduce confusion. Setting the environment for child processes through Command is probably enough. Using env::set_var for setting global state within the same process seems a bit of an anti-pattern.

**Comment by joshtriplett** [comment]
@jhpratt Do you have the bandwidth to update your copy of the branch for `#[deprecated_safe]` and get it merged?

If not, does someone else have that bandwidth?

**Comment by jhpratt** [comment]
At the moment, I do not. I haven't even looked at the code — all I did was blindly copy the branch to avoid it being lost to time. For whoever ends up working on it, you'll almost certainly want to rebase before continuing work.

**Comment by evanj** [comment]
I ran into the Go version of this problem (see https://github.com/golang/go/issues/63567 ). I think realistically, this is a small problem, and the solutions described in this thread seem reasonable. However, I am unreasonably annoyed that this problem with the C standard has been known for decades, and we can't get that fixed. I think ultimately we need the C standard to adopt a solution.

My unrealistic idea: I wish there was a way that Rust, Go and other languages that care about this (e.g. Julia? https://github.com/JuliaLang/julia/issues/34726) could share a "safe" C implementation, as a way to either route around the C standard, and encourage the C standard to evolve a solution. I'm mentioning it in case others think this is interesting, because I'm annoyed enough to spend time on this.

**Comment by tarcieri** [comment]
@evanj there was some discussion of trying to improve the C API on IRLO: https://internals.rust-lang.org/t/thread-safe-environment-variable-mutation/17817

**Comment by RalfJung** [comment]
I think you are very reasonably annoyed. ;)

However, where would such a shared API go? Realistically I think we'd have to convince glibc, or else we'd all have to add a new dependency to all our programs and that's just not going to fly. Maybe if people from multiple languages make a joint effort, that increases our chances of convincing glibc... Not sure.

**Comment by comex** [comment]
I was wondering if there was some way to implement a tiny library such that
- any executable or dylib could *statically* link the library, but
- if multiple copies of the library ended up in the same process (because multiple images each had their own statically linked copy), then the copies would somehow detect each other and agree on a single piece of shared state.

By being statically linked, the library could be included with the Rust standard library – and any other language runtimes that wanted to participate – without adding a new user-visible dependency.  By sharing state, the library copies could all have a single source of truth for the current process's "thread-safe environment".

Sharing state doesn't seem that hard: on Unix, a global weak symbol would work in almost all cases.  But there's at least one case where global weak symbols end up not actually being globally unique: when a dylib is linked with linker options that limit its exported symbols list.  Is that a deal-breaker?  Maybe not?

I also don't know how this would work on Windows, which doesn't support global weak symbols.  Maybe it would suffice to *not* invent a new source of truth there and just use the existing thread-safe `GetEnvironmentVariable`/`SetEnvironmentVariable`.

…This is all probably a pipe dream though.

**Comment by ChrisDenton** [comment]
It wouldn't be needed on Windows because, as you say, modifying then environment is already thread safe.

**Comment by briansmith** [comment]
This issue is now over 2 years old (and we've known about it a lot longer than that). It took glibc 3 years from when `getrandom` was added to Linux to add support, and even now (9 years later) we still can't rely on `getrandom` without fallbacks. An interoperable memory-safe replacement for these APIs is realistically a decade-long (most likely, multi-decade) endeavor where most of the work would need to happen in the C/POSIX standardization and implementation space, not here. I think it would be great if somebody were to do that, but I don't think we should leave these APIs as they are for years longer.

I would also say that I don't have much hope that there would be a `deprecated_safe` in Rust in a timely manner. Seems like we should pursue solutions that don't involve inventing new infrastructure, such as what was suggested long ago: create new `unsafe` APis with new names that do the same, mark the existing APIs `#[deprecated]` and have them forward to the new `unsafe` functions, and have the deprecation note for each function reference its replacement.

**Comment by jhpratt** [comment]
`#[deprecated_safe]` would be great, but the person who was working on the implementation deleted their GitHub account; I have the branches backed up on my fork of rust-lang/rust.

There is a PR open (#94619) to add unsafe APIs with new names. I have no problem updating it to resolve the conflicts and whatever else may be necessary. It's currently marked as blocked on whether `#[deprecated_safe]` was going to happen, but that doesn't seem likely in the near future.
